cadence/contracts/bridge/FlowEVMBridge.cdc

```diff
@@ -107,23 +112,32 @@ access(all) contract FlowEVMBridge {
     /// @param to: The NFT recipient in FlowEVM
     /// @param tollFee: The fee paid for bridging
     ///
-    access(contract) fun bridgeNFTToEVM(token: @{NonFungibleToken.NFT}, to: EVM.EVMAddress, tollFee: @{FungibleToken.Vault}) {
+    access(all) fun bridgeNFTToEVM(
+        token: @{NonFungibleToken.NFT},
+        to: EVM.EVMAddress,
+        tollFee: @FlowToken.Vault
+    ): @FlowToken.Vault {
         pre {
-            FlowEVMBridgeUtils.validateFee(&tollFee, onboarding: false): "Invalid fee paid"
             !token.isInstance(Type<@{FungibleToken.Vault}>()): "Mixed asset types are not yet supported"
             self.typeRequiresOnboarding(token.getType()) == false: "NFT must first be onboarded"
         }
         let tokenType = token.getType()
-        let tokenID = token.getID()
-        let evmID = CrossVMNFT.getEVMID(from: &token) ?? UInt256(token.getID())
+        let tokenID = token.id
+        let evmID = CrossVMNFT.getEVMID(from: &token as &{NonFungibleToken.NFT}) ?? UInt256(token.id)
         // TODO: Enhance metadata handling on briding - URI should provide serialized JSON metadata when requested
+        //      Consider serialization util to handle this
         // Grab the URI from the NFT
         var uri: String = ""
         if let display = token.resolveView(Type<MetadataViews.Display>()) as! MetadataViews.Display? {
             uri = display.thumbnail.uri()
         }
-        FlowEVMBridgeUtils.depositTollFee(<-tollFee)
-        FlowEVMBridgeNFTEscrow.lockNFT(<-token)
+
+        // Lock the NFT & calculate the storage used by the NFT
+        let storageUsed = FlowEVMBridgeNFTEscrow.lockNFT(<-token)
+        // Calculate the bridge fee on current rates, withdraw from provided Vault and deposit to self
+        let feeAmount = FlowEVMUtils.calculateBridgeFee(used: storageUsed, includeBase: true)
+        assert(tollFee.balance >= feeAmount, message: "Insufficient fee paid to bridge this NFT")
```

Comment:
Dete had suggested we use a capability to grab the balance, and don't take in an actual vault.

I like that for one reason that we don't have to return the excess, and simply just don't keep reference to that capability, so it is still a one-time thing. 

Also, it'd be nice if we had a separate function to estimate the fee amount before the depositToEVM call comes in, so we can get the exact tollFee needed. I think returning the surplus toll fee feels unnecessary with both or one of the above ^

Reply:
I remember talking about using something like `ScopedProvider`, but when I got into the implementation, it felt like it added a ton of weight to the bridging transaction which we can get into in real-time review.
Basically, in the bridging transaction, I have to check for an existing restricted access resource. If it's not there, I issue a Capability, create & save a new resource at a unique path with a set withdrawal limit. Then I grab a reference to it to pass in with my bridge request. Or I can just withdraw the max and deposit the returned surplus. The latter feels a lot simpler.